### PR TITLE
[Bugfix] Fix NPE in ReplicaShardAllocator (#13993)

### LINKED ISSUE #13993: [BUG] NPE in ReplicaShardBatchAllocator during node drops
### Describe the bug

```
[2024-05-28T14:34:01,627][ERROR][o.o.c.c.Coordinator      ] [66a99e3966c42f12f229f8b22da6d074] unexpected failure during [node-left]
java.lang.NullPointerException: Cannot invoke "org.opensearch.cluster.routing.ShardRouting.currentNodeId()" because "primaryShard" is null
        at org.opensearch.gateway.ReplicaShardAllocator.cancelExistingRecoveryForBetterMatch(ReplicaShardAllocator.java:108)
        at org.opensearch.gateway.ReplicaShardBatchAllocator.processExistingRecoveries(ReplicaShardBatchAllocator.java:73)
        at org.opensearch.gateway.ShardsBatchGatewayAllocator.afterPrimariesBeforeReplicas(ShardsBatchGatewayAllocator.java:198)
        at org.opensearch.cluster.routing.allocation.AllocationService.allocateAllUnassignedShards(AllocationService.java:639)
        at org.opensearch.cluster.routing.allocation.AllocationService.allocateExistingUnassignedShards(AllocationService.java:609)
        at org.opensearch.cluster.routing.allocation.AllocationService.reroute(AllocationService.java:588)
        at org.opensearch.cluster.routing.allocation.AllocationService.reroute(AllocationService.java:529)
        at org.opensearch.cluster.routing.allocation.AllocationService.disassociateDeadNodes(AllocationService.java:339)
        at org.opensearch.cluster.coordination.NodeRemovalClusterStateTaskExecutor.getTaskClusterTasksResult(NodeRemovalClusterStateTaskExecutor.java:123)
        at org.opensearch.cluster.coordination.NodeRemovalClusterStateTaskExecutor.execute(NodeRemovalClusterStateTaskExecutor.java:113)
        at org.opensearch.cluster.service.MasterService.executeTasks(MasterService.java:882)
        at org.opensearch.cluster.service.MasterService.calculateTaskOutputs(MasterService.java:434)
        at org.opensearch.cluster.service.MasterService.runTasks(MasterService.java:301)
        at org.opensearch.cluster.service.MasterService$Batcher.run(MasterService.java:212)
        at org.opensearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:209)
        at org.opensearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:247)
        at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:863)
        at org.opensearch.common.util.concurrent.PrioritizedOpenSearchThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedOpenSearchThreadPoolExecutor.java:283)
        at org.opensearch.common.util.concurrent.PrioritizedOpenSearchThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedOpenSearchThreadPoolExecutor.java:246)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:840)
```

### Related component

Cluster Manager

### To Reproduce

Perform all node drops from the cluster.

### Expected behavior

Handle the NPE.

### Additional Details

**Plugins**
Please list all plugins currently enabled.

**Screenshots**

**Host/Environment (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by DaniilRoman**
### Description
The PR fixes the NPE in ReplicaShardAllocator when we can't get an active primary shard.

### Related Issues
Resolves #13993

### Check List
- [x] Functionality includes testing.

### Flaky test
- https://github.com/opensearch-project/OpenSearch/issues/13503

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java [hunk 1]
```diff
@@ -100,7 +100,10 @@ protected Runnable cancelExistingRecoveryForBetterMatch(
         Metadata metadata = allocation.metadata();
         RoutingNodes routingNodes = allocation.routingNodes();
         ShardRouting primaryShard = allocation.routingNodes().activePrimary(shard.shardId());
-        assert primaryShard != null : "the replica shard can be allocated on at least one node, so there must be an active primary";
+        if (primaryShard == null) {
+            logger.trace("{}: no active primary shard found or allocated, letting actual allocation figure it out", shard);
+            return null;
+        }
```
Comment (DaniilRoman) [inline server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:106]: Above we already return `null` for the same NPE check
Screenshot 2024-06-18 at 12 36 26

**Comment by DaniilRoman** [comment]
Hi team 👋 
Is there anything I need to improve to get an approval? 🙏

**Review by dblock — APPROVED** [review]
This looks good, but I am not super familiar with this part of the code, maybe @andrross or @mch2 can double-check this?

**Review by dblock — APPROVED** [review]
Looks innocent enough to merge.

**Comment by dblock** [comment]
Thanks for fixing this @DaniilRoman, backport is running.

File: server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java [hunk 2]
```diff
@@ -100,7 +100,10 @@ protected Runnable cancelExistingRecoveryForBetterMatch(
         Metadata metadata = allocation.metadata();
         RoutingNodes routingNodes = allocation.routingNodes();
         ShardRouting primaryShard = allocation.routingNodes().activePrimary(shard.shardId());
-        assert primaryShard != null : "the replica shard can be allocated on at least one node, so there must be an active primary";
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:103]: @DaniilRoman do we know why this assertion had to be removed. @dblock In general I am not of fan of removing assertions, they exist in the system for a reason, we shouldn't let our guards down.

File: server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java (hunk 1 above)
Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:106]: As I understand one is about the primary shard being in an assigned state the other is whether or not the physical data it hosts is present. Both IMO are different

File: server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java (hunk 2 above)
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:103]: @Bukhtawar the assertion has been replaced by an explicit null check, returning `null` instead of throwing an exception.  Either way the code will not proceed past this point. There is a previous `return null` at line 98, so it seems null is already an expected return value.

Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:103]: @dbwiddis If I am understanding this correct there are two things
1. the method `cancelExistingRecoveryForBetterMatch` is expected to return a `Runnable` to denote some work to cancel existing recovery. A `null` as a method denotes no work is supposed to be done for cancelling. 
2. the invariant that says for a given shard(supposedly replica) it's corresponding `primary` should be non-null, since the allocator is supposed to kick in to assign a `replica` copy 
So there are two different things that exist and equating them IMO is incorrect. Removing the assertion essentially says at this point of execution a primary shard can be null when the system was designed ensuring that it shouldn't?

Comment (dbwiddis) [reply server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:103]: > So there are two different things that exist and equating them IMO is incorrect. Removing the assertion essentially says at this point of execution a primary shard can be null when the system was designed ensuring that it shouldn't?

I haven't dug into the logic, but from the face of it, the previous situation is that the assertion causes an exception, and the exception isn't handled.

I've no objection to reverting this change if you think the exception should be caught and addressed elsewhere.

Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:103]: >I haven't dug into the logic, but from the face of it, the previous situation is that the assertion causes an exception, and the exception isn't handled.

Unfortunately assertions don't run outside tests and the stack trace isn't an assertion tripping. Even if an assertion is failing the right thing to do is investigate why it is failing rather than burying the problem

Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:103]: >I've no objection to reverting this change if you think the exception should be caught and addressed elsewhere.

Yes please let's revert this commit
cc: @dblock

Comment (dbwiddis) [reply server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:103]: > Unfortunately assertions don't run outside tests

@Bukhtawar that's not the case with the `assert` keyword as it's used here.  It throws an `AssertionError`.

As an `Error` it's not normally caught by exception catch blocks unless you're specifically looking for it, and by the traditional interpretation of `Error` means that it's an unrecoverable situation and the entire JVM should be shut down.

Is it your opinion that in the case there is no primary shard, that the node should restart?

Comment (dbwiddis) [reply server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:103]: @Bukhtawar @dblock  @SwethaGuptha @shiv0408 looking closer at the original issue (#13993) it produces an NPE at line 108, so it seems the stack trace on that issue is on a version prior to the one that added this `assert` statement.  The assertion was added in #10254 so it looks like the stack trace in #13993 was generated from the code prior to that other PR: https://github.com/opensearch-project/OpenSearch/blob/12115d1ad1feca522373220f5bef367f0c8008aa/server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java#L108 (CC:)

So generally my thoughts:
 - The NPE bug was already "fixed" by #13993 before the issue was filed
 - The "fix" replaced an NPE with an AssertionError instead which is changed behavior which might itself cause a node drop (except the issue indicates this is happening during a node drop)
 - I think we all need a bit more insight into when exactly this situation occurs and whether throwing an `Error` makes things worse

Comment (shiv0408) [reply server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:103]: @dbwiddis #10254 didn't "fix" or modify the behavior of this method, I had just refactored the core logic to a protected method which could be called from child batch allocator class. @SwethaGuptha can you please provide more details on which version of code you were running when you encountered this error.

Comment (dbwiddis) [reply server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:103]: > #10254 didn't "fix" or modify the behavior of this method, I had just refactored ...

So I see looking again... sorry. The first time I was just looking for a line number.  But now I'm really confused.  In order for an NPE on line 108 on null `primaryShard` it would have had to pass the assertions on 106 and 107.
https://github.com/opensearch-project/OpenSearch/blob/12115d1ad1feca522373220f5bef367f0c8008aa/server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java#L106-L108

Is this just undefined/unpredictable behavior during JVM shutdown (node drop)?

In any case, seems a revert of this PR is the right action at this point.

Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:103]: > @Bukhtawar that's not the case with the assert keyword as it's used here. It throws an AssertionError.
As an Error it's not normally caught by exception catch blocks unless you're specifically looking for it, and by the traditional interpretation of Error means that it's an unrecoverable situation and the entire JVM should be shut down.

Sorry I don't think this is the correct understanding of how `assert` works. AssertionErrors are only configured for tests using a `-ea` JVM flag and in this case I don't think an `AssertionError` was triggered. Please feel free to read up more on this [here](https://docs.oracle.com/javase/8/docs/technotes/guides/language/assert.html#:~:text=An%20assertion%20is%20a%20statement,than%20the%20speed%20of%20light.)

>In any case, seems a revert of this PR is the right action at this point

Thanks for your understanding.

Comment (dbwiddis) [reply server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:103]: Got it, and thanks for educating me.

So wouldn't an appropriate fix for the NPE then to both put the `assert` back _and_ null check to avoid the exception?  That seems odd but I've never really used assertions like this so at this point I really don't know the best approach.

Comment (andrross) [reply server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:103]: > So wouldn't an appropriate fix for the NPE then to both put the assert back and null check to avoid the exception?

@dbwiddis The appropriate fix is to figure out why the invariant is being violated and fix _that_ bug. i.e. if no primary exists, then there can be no recovery and this method should not be called. The previous behavior of having the code fail with an NPE is preferable to ignoring a situation that is not intended to be possible.

Slightly off topic, but in general I really dislike using the `assert` keyword. Instead of `assert primaryShard != null : "message"` I almost always prefer something like `Objects.requireNonNull(primaryShard, "message")` so that it deterministically fails in all situations. Only if the code is super performance critical and if the assertion check is computationally intensive do I think the `assert` keyword makes sense.

Comment (dbwiddis) [reply server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:103]: > The appropriate fix is to figure out why the invariant is being violated and fix that bug. 

@andrross In general I agree. 

In reality, I read the issue reporting this bug and it indicates it's happening during a node drop.  Which means the JVM is shutting down for \<insert reason here\>.  Which means undefined/unpredictable behavior.

TLDR: I'm not sure there is a bug, and the symptom may be a result of a JVM being shut down causing a node drop as a result of some totally unrelated thing.

There's not enough information in the issue to infer more.
